library/core/src/gc.rs

```diff
@@ -90,9 +97,5 @@ impl<T: ?Sized> DerefMut for FinalizeUnchecked<T> {
 #[cfg(not(bootstrap))]
 unsafe impl<T> FinalizerSafe for FinalizeUnchecked<T> {}
 
-#[unstable(feature = "gc", issue = "none")]
-#[cfg_attr(not(test), rustc_diagnostic_item = "ReferenceFree")]
-pub auto trait ReferenceFree {}
-
-impl<T> !ReferenceFree for &T {}
-impl<T> !ReferenceFree for &mut T {}
+#[cfg(not(bootstrap))]
+unsafe impl<T> ReferenceFree for FinalizeUnchecked<T> {}
```

Comment:
Can you explain why this is the right thing to do? I don't know that it's the *wrong* thing to do, but I'm not sure either way!

Reply:
Sure! So `FinalizeUnchecked<T>` is our _escape hatch_ struct. It is unsafe to construct one so users should only be using it if they know what they're doing anyway. The purpose of `FinalizeUnchecked` is to inform FSA that it is being too conservative and that `T`'s finalizer is in fact safe. Without implementing `ReferenceFree`, `FinalizeUnchecked<T>` types will still cause compile-time errors when `T` contains a reference -- even if that reference is never used.

Reply:
OK, I can buy that.